imsim/diffraction_fft.py

```diff
@@ -0,0 +1,226 @@
+    pixel_length: float,
+    psf_size: int,
+):
+    """2d convolve, leaving image dimension invariant.
+    image: Brightness pixel data
+    rottelpos: Telescope rotation angle [rad]
+    exptime: Exposure time (used to determine the spike width)
+    latitude: Geographic latitude of observators
+    azimuth: Azimuth of the telescope
+    altitude: Altitude of the telescope
+    full_well: Threshold value for saturated pixels
+    pixel_length: Scaling factor: The length of 1px in `spike_profile`
+    psf_size: Size of the PSF (width and height) which will be layed over saturated pixels.
+    """
+
+    psf_w = psf_h = psf_size  # neglect the outer region (~1/r^2)
+    sin_cos = np.empty(2)
+    e_z_0, e_z = prepare_e_z(latitude)
+    e_focal = e_equatorial(latitude=latitude, azimuth=azimuth, altitude=altitude)
+    field_rotation_sin_cos(e_z_0, e_z, e_focal, exptime, sin_cos)
+    d_alpha = np.arctan2(sin_cos[1], sin_cos[0])
+    rottelpos = np.pi/4. - rottelpos
+    psf = prepare_psf_field_rotation(
+        psf_w,
+        psf_h,
+        resolution=pixel_length,
+        alpha=rottelpos,
+        d_alpha=d_alpha,
+    )
+    for region_row, region_col in saturated_clusters(image, full_well):
+        img_region = image[region_row, region_col]
+        diffracted = scipy.signal.convolve2d(psf, img_region, mode="full")
+        img_region[()] = 0.0
+        add_image(
+            image,
+            diffracted,
+            row=region_row.start - psf_w,
+            col=region_col.start - psf_h,
+        )
+
+
+def add_image(image, overlay, row, col) -> None:
+    """Add overlay to image[row:,col:] cropping overlay when needed."""
+    row_start = max(0, row)
+    row_end = min(row + overlay.shape[0], image.shape[0])
+    col_start = max(0, col)
+    col_end = min(col + overlay.shape[1], image.shape[1])
+    image[row_start:row_end, col_start:col_end] += overlay[
+        max(0, -row) : min(overlay.shape[0], image.shape[0] - row),
+        max(0, -col) : min(overlay.shape[1], image.shape[1] - col),
+    ]
+
+
+def saturated_clusters(image, full_well: float):
+    """Detect clusters of saturated pixels ( > full_well) in an image.
+    Returns a list of 2-tuples of slices (pixel ranges (x and y) containing one cluster).
+    """
+    (w, h) = image.shape
+    xy = np.c_[
+        np.meshgrid(
+            np.arange(w),
+            np.arange(h),
+            indexing="ij",
+            copy=False,
+        )
+    ]
+    pixel_coordinates = xy[:, image > full_well]
+    regions = []
+    for i, j in pixel_coordinates.T:
+        neighbor_indices = [n for n, r in enumerate(regions) if dist_2d(r, (i, j)) <= 1]
+        pix_region = (slice(i, i + 1), slice(j, j + 1))
+        if not neighbor_indices:
+            regions.append(pix_region)
+            continue
+        neighbor_regions = [regions.pop(n) for n in reversed(neighbor_indices)]
+        regions.append(merge_regions(neighbor_regions + [pix_region]))
+    return regions
```

Comment:
This seems like a lot of work to avoid doing this on any pixels below threshold.  Since convolve2d wants rectangular arrays, why not just return a single range that encompasses all the central pixels.  It's not like it's wrong to include lower surface brightness pixels near the center.  And I suspect it would be faster to just have a single call to convolve2d, even though it would technically be doing more work.